### PR TITLE
add blog: how to add middleware

### PR DESCRIPTION
**Pull request by miekg**

### CODE REVIEW

**Comment by miekg** [comment]
@johnbelamaric this is the blog PR

File: content/post/how-to-add-middleware.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
++++
+date = "2017-02-20T20:46:48Z"
+slug = "How to add middleware to CoreDNS; a tutorial"
+tags = ["middleware", "howto", "tutorial"]
+title = "How to Add Middleware to CoreDNS"
++++
+
+CoreDNS is a DNS server that chains middleware. Middleware is defined as a method `ServeDNS()` that
+gets a request and either responds to the client or passes it on to the next middleware. If none of
+the middlewares handle the request a default response of SERVFAIL is returned.
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/middlewares/middlware/ (it is its own plural, like "software")

File: content/post/how-to-add-middleware.md [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+middleware which is a CoreDNS middleware and loaded by default if nothing is specified.
+
+> Note all the code examples here are in Go because CoreDNS is written in the Go language.
+
+First question should be: "What should my middleware do?". In the case of the *whoami* middleware
+its purpose is to echo back the client's IP (IPv4 or IPv6), the transport used ("tcp" or "udp") and
+the request port number.
+
+Next question is: "What is the name of this new middleware?" Try to find a short, descriptive name
+for this. In case this, we already had a (good) name: *whoami*.
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/case this/this case/

File: content/post/how-to-add-middleware.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+## Middleware
+
+A middleware consists out of an number of parts:
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/ out//

Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/an/a/

File: content/post/how-to-add-middleware.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+1. The registration of the middelware
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/middelware/middleware/

File: content/post/how-to-add-middleware.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+2. The setup function that parses the *whois* directive and
+    possible option from the Corefile
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/option/options/

File: content/post/how-to-add-middleware.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+3. The `serveDNS()` and `Name()` methods
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: Capitalize `ServeDNS`

File: content/post/how-to-add-middleware.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+4. Hooking it up
+5. Using it
+
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: 4 and 5 aren't really "parts", they are things that you have to do. Maybe reword this a little?

File: content/post/how-to-add-middleware.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+		ServerType: "dns",
+		Action:     setupWhoami,
+	})
+}
+~~~
+First you see that we're using `caddy.RegisterPlugin` which makes perfectly sense, because
+CoreDNS uses the [Caddy](https://caddyserver.com) server framework. The first argument is the name
+of the directive as used in the Corefile.
+
+The `ServerType` is "dns" because CoreDNS is of that servertype. The `Action` is the name of the
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/is of that servertype/is a DNS server; Caddy supports HTTP servers as well./

File: content/post/how-to-add-middleware.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+~~~
+// ServeDNS implements the middleware.Handler interface.
+func (wh Whoami) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+~~~
+As seen the function gets a couple of parameters of which `w` is the client side: writing to `w`
+returns a response to the client. As you will see below all interesting properties of the client
+connection can be retrieved from the
+[dns.ResponseWriter](https://godoc.org/github.com/miekg/dns#ResponseWriter). `r` is the incoming
+query. The `ServeDNS` returns a integer and/or an error. The values the integer can take are
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/returns/method returns/
s/a integer/an integer/

File: content/post/how-to-add-middleware.md [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+the DNS RCODEs, dns.ServerFailure, dns.NotImplemented, dns.Success, etc. etc. A successful return
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/etc. etc./etc./

File: content/post/how-to-add-middleware.md [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+value indicates the middleware has written to the client.
+
+`request.Request` is a helper struct that abstracts and cache some client-side properties, like
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/cache /caches /

File: content/post/how-to-add-middleware.md [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+            Rrtype: dns.TypeA, Class: state.QClass()}
+    rr.(*dns.A).A = net.ParseIP(ip).To4()
+case 2:
+    rr = new(dns.AAAA)
+    rr.(*dns.AAAA).Hdr = dns.RR_Header{Name: state.QName(),
+            Rrtype: dns.TypeAAAA, Class: state.QClass()}
+    rr.(*dns.AAAA).AAAA = net.ParseIP(ip)
+}
+~~~
+`IP()` returns the IP address of the client. `Family()` returns the transport used. Depending on the
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: When you say "transport" I think UDP/TCP. I think what you mean is something like "returns the IP address version".

File: content/post/how-to-add-middleware.md [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+return 0, nil
+~~~
+
+First we add the two created resource records (`rr` and `srv`) to the additional section of the
+answer message: `a.Extra`.
+Then we call `state.SizeOnDo(a)` which will set the correct EDNS0 record on the message and handle
+setting the truncation bit). Then,
+finally, we call the `WriteMsg` method on `w` which writes the message back to the client.
+We indicate a successful (even though it might have failed - we didn't check the return value of
+`WriteMsg`) write be returning 0 and `nil`.
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/be/by/

File: content/post/how-to-add-middleware.md [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+# 4. Hooking it up
+
+Next we need to tell CoreDNS to compile and use this new middleware. Adding a middleware was
+recently simplified and only consists out of editing `middleware.cfg` and adding the line:
+
+~~~
+210:whoami:whoami
+~~~
+The initial number is used for sorting the middlware (more on that below), then the name of the
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/middlware/middleware/

File: content/post/how-to-add-middleware.md [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+plugin as used in the registration and then the package inside the CoreDNS middleware directory.
+
+Each middleware has a place in the list of all other middlewares. For instance the caching
+or the metrics middleware needs to come early, so that it can "see" the query and
+response and do cachy or metricsy things with it. A *whoami* middleware is not that special and can
+be placed relatively late in the list (hence the high number).
+
+Now do a `go gen && go build` to get a `coredns` binary that can be used with our shiny new middleware.
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: Or just "make"...

File: content/post/how-to-add-middleware.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+# 5. Using it
+
+Write a Corefile:
+
+~~~ txt
+. {
+    whoami
+}
+~~~
+In words this says: be authortitative for the root `.` and below, meaning all posible queries will
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/posible/possible/

File: content/post/how-to-add-middleware.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+
+Let's try using TCP:
+~~~ sh
+% dig +nocmd @localhost a example.org -p1053 +noall +additional +tcp
+example.org.		0	IN	A	127.0.0.1
+_tcp.example.org.	0	IN	SRV	0 0 33435 .
+~~~
+Yes, it correctly sees we've used TCP this time (and of course a different port).
+
+There are already a lot of middlewares in CoreDNS. New, exiting once, are always welcome. So if you
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/exiting/exciting/

**Comment by miekg** [comment]
Thanks for the review. Acting upon all comments. PTAL.

File: content/post/how-to-add-middleware.md [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+
+Next question is: "What is the name of this new middleware?" Try to find a short, descriptive name
+for this. In this case, we already had a (good) name: *whoami*.
+
+## Middleware
+
+A middleware consists of a number of parts:
+
+1. The registration of the middleware
+2. The setup function that parses the *whois* directive and
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/*whois*/*whoami*/

Although, maybe it should just say "parses the registered directive" since the rest of this list applies to all middleware not just *whoami*.

File: content/post/how-to-add-middleware.md [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+query. The `ServeDNS` method returns an integer and/or an error. The values the integer can take are
+the DNS RCODEs, dns.RcodeServerFailure, dns.RcodeNotImplemented, dns.RcodeSuccess, etc..
+A successful return value indicates the middleware has written to the client.
+
+`request.Request` is a helper struct that abstracts and caches some client-side properties, like
+EDNS0 records and the DNSSEC OK bit.
+
+Next we setup the reply message:
+~~~ go
+a := new(dns.Msg)
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: Why new(dns.Msg) and not &dns.Msg? I am relatively new to go...

File: content/post/how-to-add-middleware.md [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+Then we call `state.SizeOnDo(a)` which will set the correct EDNS0 record on the message and handle
+setting the truncation bit). Then,
+finally, we call the `WriteMsg` method on `w` which writes the message back to the client.
+We indicate a successful (even though it might have failed - we didn't check the return value of
+`WriteMsg`) write by returning 0 and `nil`.
+
+# 4. Hooking it up
+
+Next we need to tell CoreDNS to compile and use this new middleware. Adding a middleware was
+recently simplified and only consists out of editing `middleware.cfg` and adding the line:
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/out of/of/

File: content/post/how-to-add-middleware.md [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+
+Let's try using TCP:
+~~~ sh
+% dig +nocmd @localhost a example.org -p1053 +noall +additional +tcp
+example.org.		0	IN	A	127.0.0.1
+_tcp.example.org.	0	IN	SRV	0 0 33435 .
+~~~
+Yes, it correctly sees we've used TCP this time (and of course a different port).
+
+There are already a lot of middlewares in CoreDNS. New, exciting once, are always welcome. So if you
```
Comment (johnbelamaric) [inline content/post/how-to-add-middleware.md]: s/once/ones/

I also would change the 'middlewares' to 'middleware'. Maybe instead of "a lot of middleware" it would get the point across to say "a lot of different middleware".

Although if you prefer the 's' it's still OK to have changed it in the other place above ("none of the middleware"), because "none" is singular.

File: content/post/how-to-add-middleware.md (hunk 19 above)
Comment (miekg) [inline content/post/how-to-add-middleware.md]: Old habit: shouldn't use `new` for new stuff, but the struct literal approach: &...{}. Removed

File: content/post/how-to-add-middleware.md (hunk 18 above)
Comment (miekg) [inline content/post/how-to-add-middleware.md]: only fixed the typo

File: content/post/how-to-add-middleware.md (hunk 21 above)
Comment (miekg) [inline content/post/how-to-add-middleware.md]: went with a lot of different middleware.

PTAL
